src/modules/onlstm/ON_LSTM.py

```diff
@@ -0,0 +1,241 @@
+                                    for i in range(len(layer_sizes) - 1)])
+        self.lockdrop = LockedDropout()
+        #self.lockdrop2= LockedDropout()
+        self.dropout = dropout
+        self.dropouti = dropouti
+        self.dropouth = dropouth
+        self.sizes = layer_sizes
+        self.embedder = embedder
+        dim = self.embedder.token_embedder_words.weight.shape
+        self.emb = nn.Embedding(dim[0], dim[1])
+        #self.hidden=self.init_hidden(batch_size)
+        self._phrase_layer = phrase_layer
+
+        self.dropoutw = dropoutw
+
+    def get_input_dim(self):
+        return self.layer_sizes[0]
+
+    def get_output_dim(self):
+        return self.layer_sizes[-1]
+
+    def init_hidden(self, bsz):
+        return [c.init_hidden(bsz) for c in self.cells]
+
+    def forward(self, input, task=None):
+        #input= torch.transpose(input, 0, 1)
+        batch_size = input.size()[1]
+        #hidden=self.hidden
+        hidden = self.init_hidden(batch_size)
+        return self.forward_actual(input, hidden)
+
+    def forward_actual(self, input, hidden):
+        abs_inp = input
+        input = embedded_dropout(self.emb, input, dropout=self.dropoutw if self.training else 0)
+        input = self.lockdrop(input, self.dropouti)
+        length, batch_size, _ = input.size()
+        if self.training:
+            for c in self.cells:
+                c.sample_masks()
+
+        prev_state = list(hidden)
+        prev_layer = input
+
+        raw_outputs = []
+        outputs = []
+        distances_forget = []
+        distances_in = []
+        for l in range(len(self.cells)):
+            curr_layer = [None] * length
+            dist = [None] * length
+            t_input = self.cells[l].ih(prev_layer)
+            for t in range(length):
+                hidden, cell, d = self.cells[l](
+                    None, prev_state[l],
+                    transformed_input=t_input[t]
+                )
+                prev_state[l] = hidden, cell  # overwritten every timestep
+                curr_layer[t] = hidden
+                dist[t] = d
+
+            prev_layer = torch.stack(curr_layer)
+            dist_cforget, dist_cin = zip(*dist)
+            dist_layer_cforget = torch.stack(dist_cforget)
+            dist_layer_cin = torch.stack(dist_cin)
+            raw_outputs.append(prev_layer)
+            if l < len(self.cells) - 1:
+                prev_layer = self.lockdrop(prev_layer, self.dropouth)
+            outputs.append(prev_layer)
+            distances_forget.append(dist_layer_cforget)
+            distances_in.append(dist_layer_cin)
+        output = prev_layer
+        output = self.lockdrop(output, self.dropout)
+        mask = abs_inp != 0
+        self.distances = torch.stack(distances_forget)
+        return output, mask
+        #return output, prev_state, raw_outputs, outputs, (torch.stack(distances_forget), torch.stack(distances_in))
+
+
+if __name__ == "__main__":
```

Comment:
Add a comment explaining this.

Reply:
Removed the main call completely as this was sort of redundant.